#[107791770] Taskee log

Habu-Kagumba opened the pull request:
#### What does this PR do?

Enable a `taskee` to view a log of **done**, **active** and **inactive** tasks.
#### How should this be manually tested?

As a taskee, login to the dashboard and click on the `Task logs` link.
#### What are the relevant pivotal tracker stories?
#107791770
#### Screenshots (if appropriate)

![](http://cdn.kagumba.com/assets/taskee_log.png)

codesword commented on app/controllers/task_managements_controller.rb:
```diff
@@ -23,6 +54,20 @@ def update
+      task_category,
+      notif_tasker,
+      notif_taskee
+    ).deliver_now if @notif == true
```
When you have situations like this `if @notif == true`, it will make sense to return early instead of at the very end. Reason been that you don't what to speed the time and resources finding taskee, tasker and task_category before discovering that a user didn't enable notifications. You can have something like this.

``` ruby
def send_email_notifications(task)
  return unless current_user.enable_notifications
  # do your things
end
```

I also noticed that you are passing a lot of arguments to `NotificationMailer.send_notifications`. This probably be a good time to use a `View Object`/,`Presenters`
.

codesword commented on app/controllers/task_managements_controller.rb:
```diff
@@ -23,6 +54,20 @@ def update
 
+  def send_email_notifications(task)
+    notif_taskee = User.find_by_id(task.taskee_id)
+    notif_tasker = User.find_by_id(task.tasker_id)
```
You can just use the `find` method directly instead of `find_by_id`. Reason been that `find_by_id` has been depreciated and `find` method finds a record by id(primary key). If you want to find by a different column, you can use `find_by` method eg `User.find_by(name: "Herbert")`

codesword commented on app/views/task_managements/index.html.erb:
```diff
@@ -1,8 +1,35 @@
+                          <% user = User.find(task.tasker_id) %>
+                          <td><%= "#{user.firstname} #{user.lastname}" %></td>
+                          <td><%= task.task_desc %></td>
+                          <% if task.status.downcase == "done" %>
```
This whole logic can be extracted into a helper method or a decorator method. Strive for cleaner html pages, pushing logic into a helper, a decorator or a view object/presenter.

codesword commented on app/controllers/taskmanagements_controller.rb:
```diff
@@ -1,94 +0,0 @@
-class TaskmanagementsController < ApplicationController
```
Good Job moving all methods in this class to `TaskManagementsController` class.

codesword commented on app/assets/stylesheets/styles/_dashboard.scss:
```diff
@@ -78,132 +78,131 @@ nav .nav-wrapper {
-        margin-top: em(7.5px);
-        border-radius: 100%;
-      }
+  .dropdown {
```
Good job cleaning up this css file.

Commit:
refactor

- add TaskManagementsPresenter to slim the task managements index view
- add helper methods for users
- update task management index.html to use a presenter

Commit:
Merge branch 'master' into ft-taskee-log-107791770

Commit:
refactor(task-presenter): Refactor the TaskManagementsPresenter

Refactor the TaskManagementsPresenter to use the built in
SimpleDelegator

Commit:
test(taskee-logs): added tests for the Taskee log feature

WIP - Add tests for the Taskee Log feature
- DRY up usage of the `log_in_with` method

[#107791770]

Commit:
test(taskee-log): Add tests for task management controller

- Update the taskee log feature
- Add tests for the task management controller

[#107791770]

lekansogunle commented on spec/factories/task_managements.rb:
```diff
@@ -0,0 +1,15 @@
+    task_id 1
+    tasker_id 1
+    taskee_id 2
+    task_desc "A task"
```
Consider using https://github.com/stympy/faker#fakerlorem for the description and other generic attributes.

Commit:
Merge branch 'master' into ft-taskee-log-107791770

lekansogunle commented on spec/factories/task_managements.rb:
```diff
@@ -1,10 +1,15 @@
     tasker_id 1
     taskee_id 2
-    amount 5000
+    task_desc "A task"
```
consider using https://github.com/stympy/faker#fakerlorem for the description. Also, use Faker for generic attributes.

lekansogunle commented on spec/features/taskee_logs_spec.rb:
```diff
@@ -0,0 +1,38 @@
+        expect(page).to have_content("inactive")
+        expect(page).to have_content("done")
+
+        expect(page.find_all("tr").size).to eql 4
```
We talked about having a general constant available for this number 4.

lekansogunle commented on spec/support/helpers.rb:
```diff
@@ -0,0 +1,25 @@
+    click_button "Sign in"
+  end
+
+  def taskee_stub
```
Nice abstraction.

Commit:
bug(taskmanagement): remove duplicate method

- remove duplicate index method

[#107791770]

Commit:
fix(taskeelog): fix taskee log specs

- move number of tasks into a variable in the support helpers

[#107791770]

Commit:
fix(TaskManagementsController): fix spec failures

- Remove un-necessary spec controller file
- Fix failing tests in the `task_managements_controller_spec` file
- User faker in task_managements factory

[#107791770]

lekansogunle commented on spec/controllers/task_managements_controller_spec.rb:
```diff
@@ -33,12 +36,15 @@
-        expect(assigns(:tasks)).to eql @user.tasks_created.to_a
+
+        assign_tasks
+        expect(@tasker.tasks_created).not_to be_blank
```
Nice work so far. It will be great to test for the exact form of the task created instead of a not blank. Something like the count of the tasks assigned.

lekansogunle commented on spec/factories/task_managements.rb:
```diff
@@ -1,10 +1,15 @@
     taskee_id 2
-    amount 5000
+    task_desc Faker::Lorem.sentence
+    amount "9000"
```
Consider using Faker::Commerce.price, https://github.com/stympy/faker#fakercommerce. There are a lot of methods you can leverage for generic parameters. Also, Enclose you call to to Faker with a {} so that it generates a new instance of the parameter every time the code calls it. Something like { Faker::Address.state }.

lekansogunle commented on spec/controllers/task_managements_controller_spec.rb:
```diff
@@ -33,12 +36,15 @@
-        @user.update_attribute(:user_type, "taskee")
         get :index
-        expect(assigns(:tasks)).to eql @user.tasks_given
+
```
We can take care of extra spacing.

Commit:
fix(taskmanagement): refactor specs and factories

- Refactor the task management controller spec and the task management
  factory.

[#107791770]

lekansogunle commented on spec/controllers/task_managements_controller_spec.rb:
```diff
@@ -33,12 +36,13 @@
         get :index
-        expect(assigns(:tasks)).to eql @user.tasks_created.to_a
+        assign_tasks
+        expect(@tasker.tasks_created.size).to eql(2)
```
In testing, you want to be clear about your assertions. @tasker was defined earlier in this test. What I think you want to test for is the controller assigned task instance matching the user tasks.

Habu-Kagumba replied:
The controller assigned task always returns `nil`.
I'm currently using

```
assigns(:tasks).to eql @tasker.tasks_created
```

Commit:
fix(taskmanagements): update specs and remove presenter

- remove the TaskManagementsPresenter

[#107791770]

Commit:
fix(taskmanagements): update specs and remove presenter

- remove the TaskManagementsPresenter

[#107791770]

Commit:
Merge branch 'ft-taskee-log-107791770' of github.com:andela/workdey into ft-taskee-log-107791770

Commit:
Merge branch 'master' into ft-taskee-log-107791770
